src/decode/lzma.rs

```diff
@@ -264,76 +302,212 @@ where
+    }
+
+    pub fn process_mode<'a, R: io::BufRead>(
+        &mut self,
+        mut rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+        mode: Mode,
+    ) -> error::Result<()> {
+        loop {
+            if let Some(unpacked_size) = self.unpacked_size {
+                if self.output.len() as u64 >= unpacked_size {
+                    break;
+                }
+            } else if match mode {
+                Mode::Run => rangecoder.is_eof()?,
+                Mode::Finish => rangecoder.is_finished_ok()?,
+            } {
+                break;
+            }
+
+            if self.tmp.position() as usize > 0 {
+                // Fill as much of the tmp buffer as possible
+                let start = self.tmp.position() as usize;
+                let bytes_read = rangecoder.read_into(&mut self.tmp.get_mut()[start..])?;
+                let bytes_read = if bytes_read < std::u64::MAX as usize {
+                    bytes_read as u64
+                } else {
+                    return Err(error::Error::LZMAError(
+                        "Failed to convert integer to u64.".to_string(),
+                    ));
+                };
+                self.tmp.set_position(self.tmp.position() + bytes_read);
+                let tmp = *self.tmp.get_ref();
+
+                // Check if we need more data to advance the decompressor
+                if Mode::Run == mode
+                    && (self.tmp.position() as usize) < MAX_REQUIRED_INPUT
+                    && self
+                        .try_process_next(
+                            &tmp[0..self.tmp.position() as usize],
+                            rangecoder.range(),
+                            rangecoder.code(),
+                        )
+                        .is_err()
+                {
+                    return Ok(());
+                }
+
+                // Run the decompressor on the tmp buffer
+                let mut tmp_reader = io::Cursor::new(&tmp[0..self.tmp.position() as usize]);
+                let mut tmp_rangecoder = rangecoder::RangeDecoder::from_parts(
+                    &mut tmp_reader,
+                    rangecoder.range(),
+                    rangecoder.code(),
+                );
+                let res = self.process_next(&mut tmp_rangecoder)?;
+
+                // Update the actual rangecoder
+                let (range, code) = tmp_rangecoder.into_parts();
+                rangecoder.set(range, code);
+
+                // Update tmp buffer
+                let end = self.tmp.position();
+                let new_len = end - tmp_reader.position();
+                self.tmp.get_mut()[0..new_len as usize]
+                    .copy_from_slice(&tmp[tmp_reader.position() as usize..end as usize]);
+                self.tmp.set_position(new_len);
+
+                if !res {
+                    break;
+                };
+            } else {
+                if (Mode::Run == mode) && (rangecoder.remaining()? < MAX_REQUIRED_INPUT) {
```

Comment:
Same comment about `if variable == Constant` being more readable than the reverse. There are a few more in the code.